sdks/java/expansion-service/src/main/java/org/apache/beam/sdk/expansion/service/JavaClassLookupTransformProvider.java

```diff
@@ -0,0 +1,449 @@
+        // Parameter name through reflection is from the class file (not through synthesizing,
+        // hence we can validate names)
+        return false;
+      }
+
+      Class<PTransform<InputT, OutputT>> parameterClass =
+          (Class<PTransform<InputT, OutputT>>) parameterFromReflection.getType();
+      Row parameterRow =
+          ExternalTransformRegistrarLoader.decodeRow(
+              parameterFromPayload.getSchema(), parameterFromPayload.getPayload());
+
+      Schema parameterSchema = null;
+      if (isPrimitiveOrWrapperOrString(parameterClass)) {
+        if (parameterRow.getFieldCount() != 1) {
+          throw new RuntimeException(
+              "Expected a row for a single primitive field but received " + parameterRow);
+        }
+        // We get the value just for validation here.
+        getPrimitiveValueFromRow(parameterRow);
+      } else {
+        try {
+          parameterSchema = SCHEMA_REGISTRY.getSchema(parameterClass);
+        } catch (NoSuchSchemaException e) {
+
+          SCHEMA_REGISTRY.registerSchemaProvider(parameterClass, new JavaFieldSchema());
+          try {
+            parameterSchema = SCHEMA_REGISTRY.getSchema(parameterClass);
+          } catch (NoSuchSchemaException e1) {
+            throw new RuntimeException(e1);
+          }
+          if (parameterSchema != null && parameterSchema.getFieldCount() == 0) {
+            throw new RuntimeException(
+                "Could not determine a valid schema for parameter class " + parameterClass);
+          }
+        }
+      }
+
+      if (parameterSchema != null && !parameterRow.getSchema().assignableTo(parameterSchema)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  private Object[] getParameterValues(
+      java.lang.reflect.Parameter[] parameters, Parameter[] payloadParameters) {
+    ArrayList<Object> parameterValues = new ArrayList<>();
+    int i = 0;
+    for (java.lang.reflect.Parameter parameter : parameters) {
+      Parameter parameterConfig = payloadParameters[i];
+      Class<?> parameterClass = parameter.getType();
+
+      Row parameterRow =
+          ExternalTransformRegistrarLoader.decodeRow(
+              parameterConfig.getSchema(), parameterConfig.getPayload());
+
+      Object parameterValue = null;
+      if (isPrimitiveOrWrapperOrString(parameterClass)) {
+        parameterValue = getPrimitiveValueFromRow(parameterRow);
+      } else {
+        SerializableFunction<Row, ?> fromRowFunc = null;
+        // SCHEMA_REGISTRY.
+        try {
+          fromRowFunc = SCHEMA_REGISTRY.getFromRowFunction(parameterClass);
+        } catch (NoSuchSchemaException e) {
+          throw new IllegalArgumentException(
+              "Could not determine the row function for class " + parameterClass, e);
+        }
+        parameterValue = fromRowFunc.apply(parameterRow);
+      }
+      parameterValues.add(parameterValue);
+      i++;
```

Comment:
nit: Use a classic for loop, its simpler to understand for what your trying to do
```suggestion
    for (int i = 0; i < parameters.length; ++i) {
      Parameter parameter = parameters[i];
      Parameter parameterConfig = payloadParameters[i];
      Class<?> parameterClass = parameter.getType();

      Row parameterRow =
          ExternalTransformRegistrarLoader.decodeRow(
              parameterConfig.getSchema(), parameterConfig.getPayload());

      Object parameterValue = null;
      if (isPrimitiveOrWrapperOrString(parameterClass)) {
        parameterValue = getPrimitiveValueFromRow(parameterRow);
      } else {
        SerializableFunction<Row, ?> fromRowFunc = null;
        // SCHEMA_REGISTRY.
        try {
          fromRowFunc = SCHEMA_REGISTRY.getFromRowFunction(parameterClass);
        } catch (NoSuchSchemaException e) {
          throw new IllegalArgumentException(
              "Could not determine the row function for class " + parameterClass, e);
        }
        parameterValue = fromRowFunc.apply(parameterRow);
      }
      parameterValues.add(parameterValue);
```